### DISCUSSION TITLE
【作业打卡】Week4. 模块四：Rust 项目基础

### DISCUSSION THREAD

**Discussion by bonnie0519** [CreatorsDAO/rust-co-learn#28] [Assignment/Tasks]
1. 你认为的错误处理逻辑应该是怎样的？
2. 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行
3. 尝试编写单元测试和集成测试

**Comment by v1xingyue** [comment]
## 你认为的错误处理逻辑应该是怎样的？

按照情况分为如下五种:

1. 类型错误，此类错误在编译初期解决
2. 可能返回None 的 用 Option 来封装。 Option 和 Some 配对使用。处理返回结果用 match
3. 如果你还想把error 传回调用者，那么使用 Result 来封装。定义完成以后，可以使用 Ok Err 返回对应的信息
4. 断言 理性防御
5. 最后一种，如果实在处理不了了，那就 panic 

## 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行

关键还是toml 文件

```rust

[package]
name = "app"
version = "0.1.0"
edition = "2021"

[workspace]
members = ["mybase","mylib"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
mylib = {path="./mylib"}

```

## 尝试编写单元测试和集成测试

1. 单元测试，在模块内部声明:

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let result = add(2, 2);
        assert_eq!(result, 4);
    }
}

```

2. 关于集成测试
集成测试一般只适用于 lib package , 同时，只能调用lib.rs 中定义的module。即作为一个整体来执行。

在 Rust 中，crate 关键字只能在库（lib）包的根模块中使用，它用于引用当前包的内容。而在 tests 目录中，测试文件是作为独立的模块执行的，无法直接访问 crate 关键字。

如果你需要强行测试，可以使用文件路径指定module 

```rust
#[path = "../src/myinfo.rs"]
mod myinfo;
```

**Comment by jiantao88** [comment]
### 你认为的错误处理逻辑应该是怎样的？
良好的错误处理逻辑的通用准则：

1. **明确和早期的错误反馈**：程序应当在出现错误的一刹那立即给出反馈，并且清晰明了地表明出现了何种错误。

2. **尽可能地提供详细信息**：错误消息应包括可能的错误原因，错误的具体位置（例如文件名，行号等），以及任何可能帮助诊断问题的其他信息。

3. **错误隔离**：在可能的情况下，错误处理逻辑应防止错误导致程序的其他部分出现问题。这可能需要使用诸如异常处理这样的技术，以便在错误发生时可以安全地清理并从错误中恢复。

4. **允许故障恢复**：在可能的情况下，应该有能力从错误中恢复，而不仅仅是停止执行。这可能需要通过重试操作，回滚到安全状态，或者给用户提供修复问题或跳过问题的选项来实现。

#### Rust处理错误的一些方式:

1. **Option 和 Result 类型**：Rust使用 Option 和 Result 类型作为函数的返回类型，以表示可能的失败。这是一种显式的错误处理方法，要求开发者在处理结果之前就必须处理可能的错误。这满足了上述的“明确和早期的错误反馈”以及“错误隔离”的准则。

2. **模式匹配**：Rust 的模式匹配可以被用于解构 Option 和 Result 类型，使得处理成功和失败的情况都非常明确。如果你不处理所有可能的结果，Rust编译器会给出警告。

3. **无异常**：Rust语言本身没有异常机制。因此，无需担心异常传播和难以跟踪的控制流，这也意味着错误必须被显式处理。

4. **panic!宏**：尽管 Rust 没有异常，但它提供了一个 panic! 宏，可以在程序遇到无法处理的错误时停止执行。panic!会导致当前线程立即停止执行，并开始回溯。这满足了"优雅的失败"的准则。

5. **显式错误类型**：Rust 鼓励你为你的 API 定义自己的错误类型，这可以让你提供关于错误的详细信息。



### 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行
[rust 工作空间project ](https://github.com/jiantao88/rust_project_ex)

### 尝试编写单元测试和集成测试

单元测试是在源代码文件中与被测试的代码一起编写的，并使用 `#[cfg(test)]` 属性标记。

```rust
// src/lib.rs

pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(add(2, 3), 5);
        assert_eq!(add(-1, 1), 0);
        assert_eq!(add(10, -5), 5);
    }
}
```

在这个示例中，`add` 函数是我们要测试的函数。我们使用 `#[cfg(test)]` 将测试代码限制在测试环境中。

同一个模块内使用 `#[test]` 属性标记一个测试函数 `test_add`。在测试函数中，我们使用 `assert_eq!` 宏来断言函数的返回值与我们期望的结果是否相等。如果断言失败，测试将失败并输出错误信息。

用 `cargo test` 命令：

```
$ cargo test
```

```rust
// tests/integration_tests.rs

use my_library::add;

#[test]
fn test_add() {
    assert_eq!(add(2, 3), 5);
    assert_eq!(add(-1, 1), 0);
    assert_eq!(add(10, -5), 5);
}
```
**集成测试**

1. 创建一个名为 `tests` 的目录，位于项目根目录下。

2. 在 `tests` 目录下创建一个新的源代码文件，以 `.rs` 为后缀，如 `integration_tests.rs`。

3. 在新创建的源代码文件中，编写测试代码。可以使用 `#[cfg(test)]` 属性将测试代码限制在测试环境中。

   ```rust
   // tests/integration_tests.rs

   #[test]
   fn test_example() {
       // 编写测试代码
       // 断言等等
   }
   ```

4. 使用 `cargo test` 命令来运行集成测试。Cargo 将自动发现并执行 `tests` 目录下的所有测试文件。

   ```
   $ cargo test
   ```

一个简单的集成测试：

```rust
// tests/integration_tests.rs

#[test]
fn test_addition() {
    assert_eq!(2 + 2, 4);
}

#[test]
fn test_subtraction() {
    assert_eq!(5 - 2, 3);
}
```

**Comment by lx7575000** [comment]
### 1. 你认为的错误处理逻辑应该是怎样的？

- 明确区分错误类型和对应错误提示，根据情况使用Result和Option配合match方法匹配对应的错误处理方法。

``` rust
let result: Result<i32, i32> = Ok(1)  // Err(404);

match result {
	Ok(value) => {
		match value {
			1 => {
				println!("成功返回值1");
			},
			2 => {
				println!("成功返回值1");
			},
			_ => {
				println!("成功默认处理方法");
			},
		}
	},
	Err(error) => {
		match error {
			404 => {
				println!("Not Found");
			},
			_ => {
				eprintln!("Error {}", error);
			}
		}
	}
}
```

- 细粒度的错误处理逻辑：尽可能在错误发生处完成对应的处理工作，减少由于防御式编程增加的冗余容错代码
- 提供错误日志：日志中尽量提供足够的信息（模块信息，错误内容等），方便开发者根据日志链路去排查解决问题。
- panic时提供提供保活或重启能力

```rust
use std::panic;

fn main() {
    restart();
}

fn restart() {
    let result = panic::catch_unwind(|| {
        // 执行可能引发panic的代码
        println!("This will be printed before panic");
        panic!("Something went wrong");
    });

    match result {
        Ok(_) => {
            // panic未发生，继续执行
            println!("The program survived panic");
        },
        Err(_) => {
            // panic发生，重新启动程序
            println!("The program panicked, restarting...");
            restart();
        },
    }
}
```

### 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行
https://github.com/lx7575000/rust-project-demo/tree/main
关键步骤：
1. 在项目根目录中声明引入的各子模块目录名，并在主模块目录中声明依赖关系

``` 
// 根目录Cargo.toml
[workspace]
members = ['entry', 'lib', 'lib2']
```
```
// entry目录Cargo.toml

[dependencies]
lib = {path="../lib"}
lib2 = {path="../lib2"}
```

2. lib库类型模块中按照实际情况逐级声明支持导出的模块

```rust
// 根据实际情况选择是否声明pub来支持外部调用
pub mod lib2_cell;
pub mod lib2_universe;
```

<img width="1144" alt="image" src="https://github.com/CreatorsDAO/rust-co-learn/assets/7943436/0ede894b-73f0-49f3-8ac4-948fef84451b">


### 3. 尝试编写单元测试和集成测试

- 单元测试是在对应模块代码文件中与被测试的模块代码一起编写的，并使用 #[cfg(test)] 属性标记。
- 集成测试是统一放在项目根路径的ltests目录下，执行`cargo test`命令调用

参照canway's life game的代码编写了webassembly模块的测试用例
<img width="779" alt="image" src="https://github.com/CreatorsDAO/rust-co-learn/assets/7943436/0a2a6d92-6c5c-4a94-8b0a-a241acd5398f">
<img width="758" alt="image" src="https://github.com/CreatorsDAO/rust-co-learn/assets/7943436/0b886581-2175-4636-9342-17c5d3b148de">

**Comment by coderonion** [comment]
## 1.你认为的错误处理逻辑应该是怎样的？

- 编译期错误：rust严格类型系统保证。
- 逻辑错误：断言，保证逻辑正确。
- 空值处理：**Option<T>**，处理**Some<T>**或None的情况，可通过match匹配或？向上传播None。
- 可恢复错误：**Result<T,E>** ，处理正确**OK<T>**或错误**Err<E>**的情况，可通过match匹配或？向上传播错误Err<E>。
- 不可恢复错误：panic!宏，当前线程终止。

## 2.以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行

https://github.com/sjinzh/rust_learn_tintin

## 3.尝试编写单元测试和集成测试

1）单元测试目标是测试某一个代码单元(一般都是函数)，验证该单元是否能按照预期进行工作。单元测试的惯例是将测试代码的模块跟待测试的正常代码放入同一个文件中，并使用 `#[cfg(test)]` 进行了标注。例如在add_pkg项目下的src/lib.rs文件中进行单元测试：

```rust
// add_pkg/src/lib.rs

pub fn add_four(a: i32) -> i32 {
    a + 4
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add_four() {
        assert_eq!(add_four(2), 6);
        assert_eq!(add_four(-2), 2);
    }
}
```

2）集成测试的代码是在一个单独的目录下的，对某一个功能或者接口进行测试。rust默认的集成测试文件夹`tests` 目录，跟 `src` 目录同级。例如在add_pkg项目下的tests/integration_test.rs文件中进行集成测试：

```rust
// add_pkg/tests/integration_test.rs

use add_pkg::add_four;

#[test]
fn integration_test_add_four() {
    assert_eq!(add_four(2), 6);
    assert_eq!(add_four(-2), 2);
}
```

**Comment by BkunS** [comment]
**1. 你认为的错误处理逻辑应该是怎样的？**
try catch。 Result + match 也不错。

**2. 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行**
见自己创建的 Leetcode问题 的 [Repo](https://github.com/BkunS/leetcode-problems-rust)

**3. 尝试编写单元测试和集成测试**
```
#[test]
fn contains_two() {
    let mut has_two = false;
    let vec = vec![0, 1, 2];
    vec.iter().for_each(|i| {
        if *i == 2 {
            has_two = true
        }
    });
    assert_eq!(true, has_two);
}
```

**Comment by laughinglky** [comment]
1、你认为的错误处理逻辑应该是怎样的？
（1）类型系统保证编译阶段就可以发现类型错误
（2）处理是否有值情况要使用Option<T>，对不同情况进行处理
（3）处理成功或失败要返回Result<T,E>，对潜在的错误进行处理
（4）assert和panic用于已知的错误情况处理

2、以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行
（1）创建项目目录：mkdir my-project
（2）进入目录my-project创建不同crate
    cargo new c1-lib --lib
    cargo new c2-lib --lib
    cargo new c3-bin --bin
（3）创建Cargo.toml文件填写成员
[workspace]
members = ["c1-lib", "c2-lib", "c3-bin"]
（4）运行项目：cargo run

**Comment by xiangaoole** [comment]
### 1. 你认为的错误处理逻辑应该是怎样的？

对于可能出现错误的方法使用Result返回结果，对于可能出现空值的方法使用Optional返回结果。谨慎使用assert和panic。




### 2. 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行

<img width="312" alt="image" src="https://github.com/CreatorsDAO/rust-co-learn/assets/32730253/b56f0726-f296-4ee6-8244-332cab3602b7">





### 3. 尝试编写单元测试和集成测试

```rust
use rand::Rng;
use std::fs::File;
use std::io::{self, Read};

fn read_file(file_path: &str) -> Result<String, io::Error> {
    let mut file = File::open(file_path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    Ok(contents)
}

pub fn get_random() -> i32 {
    return rand::thread_rng().gen_range(0..10);
}

#[cfg(test)]

mod tests {
    use super::*;

    #[test]
    fn test_random() {
        let r = get_random();
        assert!(r > 0 && r < 10)
    }

    #[test]
    fn test_result() -> Result<(),
     String> {
        match read_file("Cargo.toml") {
            Ok(contents) => {
                print!("read files: {}" , contents);
                Ok(())
            }
            Err(error) => {
                print!("read files error: {}" , error);
                Err(String::from("read files error"))
            }
        }
    }
}

```

**Comment by Insomnia501** [comment]
### 你认为的错误处理逻辑应该是怎样的？

1. 强类型系统，保证类型不出错
2. 可能无值的情况使用option处理
3. 可能error的情况使用result处理
5. 用assert来做一些确定性检查，及时中断不符合预期的逻辑
6. panic处理恐慌（如果逻辑认为某个情况一定不会出现时，用panic处理）

### 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行
略

### 尝试编写单元测试和集成测试
用于测试一个矩形是否是正方形
```
pub struct Point {
    x: f64,
    y: f64,
}

pub struct Rectangle {
    top_left: Point,
    bottom_right: Point,
}

impl Rectangle {
    pub fn is_square(&self) -> bool {
        let width = self.bottom_right.x - self.top_left.x;
        let height = self.top_left.y - self.bottom_right.y;
        return (width - height).abs() < std::f64::EPSILON;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_square() {
        let rect1 = Rectangle {
            top_left: Point { x: 1.0, y: 4.0 },
            bottom_right: Point { x: 4.0, y: 1.0 },
        };
        assert!(rect1.is_square());

        let rect2 = Rectangle {
            top_left: Point { x: 1.0, y: 5.0 },
            bottom_right: Point { x: 4.0, y: 1.0 },
        };
        assert!(!rect2.is_square());
    }
}

```

**Comment by dashjay** [comment]
## 你认为的错误处理逻辑应该是怎样的？
首先，错误需要一个良好的定义，rust 的枚举值可以支持多种类型，是一种不错的定义错误的方式，方便传递一些值来增加一些额外的信息，例如：
```
pub enum ErrorKind{
    FileNotFound(String),
    PermissionDenied(i32),
    IOError(String),
}
```
以上三种错误类型上确定了，并且额外的带有一些信息：
- FileNotFound 携带了一个 String 外部的用户知道究竟是哪一个文件找不到。
- PermissionDenied 携带了一个 i32 可能用于储存传入的权限，到底是什么权限不被允许
- IOError 携带了一个 String 可能把上层的网络 IO Error 直接存储，方便暴露出来。

以上的实现谈不上多好，只是一些初级的简单的实现
Rust 中错误处理应当合理使用 trait From 和 Option Result 等类型，搭配 `?` 可以简化大量的代码

## 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行

这是 https://github.com/dashjay/bytestack 最近编写的一个项目，内部使用了一些 workspace 项目组织，欢迎参观批评。

## 尝试编写单元测试和集成测试 

上方的 repo 中包含部分测试代码，算是打卡交作业了

**Comment by sketch2000** [comment]
1. 你认为的错误处理逻辑应该是怎样的？

   - 程序应当在出现错误的第一时间给出反馈
   - 尽可能提供详细的错误信息，方便根据这些信息排查问题
   - 提供程序继续运行的机制，有时候业务上允许错误发生

2. 以一个工作空间的形式创建一个项目，并在其下创建多个模块，并把它们关联起来，使之能够正常编译运行

   一个项目跟项目的Cargo.toml文件

   ```rust
   [workspace]
   members = [
       "abi",
       "service",
   ]
   
   [workspace.dependencies]
   anyhow = "1.0.71"
   axum = {version="0.6.18",features=["tokio","json"] }
   dotenv = "0.15.0"
   async-trait="0.1.71"
   tracing = "0.1.37"
   tokio = { version = "1.5.0", features = ["full"] }
   thiserror = "1"
   sqlx = {version = "0.7.0", features = ["postgres", "runtime-tokio-rustls", "uuid","macros","chrono"] }
   sqlb = "0.3.3"
   serde = {version = "1", features = ["derive"] }
   serde_yaml = "0.9"
   ```

3. 尝试编写单元测试和集成测试

   ```rust
   // 一个简单单元测试样例
   #[cfg(test)]
   mod tests {
     assert!(sum(1,2), 3);
   }
   ```

   单元测试一般写在与函数同文件内，

   集成测试一般写在项目根路径下，通过执行`cargo test`运行

**Comment by walle233** [comment]
1. 我认为Rust的错误处理逻辑应该是通过Result类型来处理可能发生的错误。Result类型有两个可能的结果: Ok或Err。当函数成功执行时，返回Ok包含函数的结果；当出现错误时，返回Err包含一个错误信息。开发者可以使用match语句或者Result的内置方法来处理返回的结果，从而决定程序的流程。
2. 要以工作空间的形式创建一个项目，并在其下创建多个模块，并关联它们，可以按照以下步骤操作：
   - 使用cargo new命令创建项目文件夹和Cargo.toml文件
   - 在Cargo.toml文件中添加[workspace]头部，并指定路径包含的子模块
   - 在项目文件夹下创建子模块文件夹，并在每个子模块文件夹中创建lib.rs文件
   - 在主模块文件夹中创建main.rs文件，并在其中使用mod关键字引入子模块
   - 在Cargo.toml文件中添加每个子模块的依赖关系
   - 使用cargo build或cargo run命令编译和运行项目
3. 在Rust中编写单元测试和集成测试可以使用内置的test模块。要编写单元测试，可以在被测试的函数上方使用#[test]属性，并在测试函数中使用assert_eq!宏或其他断言方法来验证函数的行为。运行测试可以使用cargo test命令。
```
pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(add(2, 3), 5);
        assert_eq!(add(-2, 2), 0);
        assert_eq!(add(-2, -3), -5);
    }
}
```

   要编写集成测试，可以在项目文件夹的tests文件夹下创建一个或多个测试文件，并使用普通的Rust代码编写测试函数。这些测试函数可以调用被测试的模块或者二进制可执行文件，并使用断言方法来验证其行为。运行集成测试可以使用cargo test命令，加上--test参数，后面跟上测试文件的名称。
```
// 集成测试文件：tests/integration_test.rs
use my_project::math::add;

#[test]
fn test_add_integration() {
    assert_eq!(add(2, 3), 5);
    assert_eq!(add(-2, 2), 0);
    assert_eq!(add(-2, -3), -5);
}

```
